### ISSUE TITLE
Q: PATH records started logging capability information

### ISSUE THREAD

**Issue by stevegrubb** [linux-audit/audit-kernel#72]
Since 4.14 kernel, it appears that we have picked up capability fields. There are only about 20 - 30 files on the whole system that have capabilities. Since capabilities are rare, they should not be logged in the general case. On my system, 0.0005% of the files have a capability. The new capability information adds 50% to the size of the old PATH records. This tradeoff just doesn't seem worth it. I don't even know why this was done. We have a AUDIT_BPRM_FCAPS record that is probably more appropriate to stuff them into so they get generated only if applicable.

**Comment by pcmoore** [comment]
@rgbriggs your the most knowledgeable on the capability auditing, I'll let you have all the fun with this one.

**Comment by rgbriggs** [comment]
On 2018-02-03 19:05, Steve Grubb wrote:
> Since 4.14 kernel, it appears that we have picked up capability
> fields. There are only about 20 - 30 files on the whole system that
> have capabilities. Since capabilities are rare, they should not be
> logged in the general case. On my system, 0.0005% of the files have a
> capability. The new capability information adds 50% to the size of the
> old PATH records. This tradeoff just doesn't seem worth it. I don't
> even know why this was done. We have a AUDIT_BPRM_FCAPS record that is
> probably more appropriate to stuff them into so they get generated
> only if applicable.

I agree it doesn't seem worth it.

So why have we spent so much time "normalizing" a number of records so
that fields don't magically appear and disappear?  I thought vanishing
fields were a hot issue and we couldn't have fields that came and went?

Again, you were explicitly copied on all of these proposed changes so
you could weigh in on them before they went upstream.  What happenned?
Why is it sometimes necessary and other times a nuisance?

It sounds like the parser library needs a rewrite for once and for all
to eliminate ordering sensitivity and default or ignore missing fields.

**Comment by stevegrubb** [comment]
Normalizing records is good. Always recording information that never changes is bad. This wastes disk space, increases processing time, consumes more memory, etc. It is OK for records to swing in and out. This is why we have so many auxiliary records. If something is rare, it goes in an auxiliary record when it can swing in and out as needed. For example, AUDIT_SOCKADDR, we do not record a sockaddr of 0.0.0.0 for each syscall even if its not applicable.

The hotbutton issues are always the same:
1) Are we recording the full information required by Common Criteria
2) Are we recording information that's not needed or never changes thus wasting disk space, increasing processing & memory, and slowing down the realtime event stream. Can this be kicked to a specific auxiliary record if sometimes important?
3) Are we reusing fields as described in the Field Dictionary rather than making up new ones
4) Are we using the exact format for the value of the field as described in the Field Dictionary?
5) Are we keeping common criteria required fields in the same order so that we can reuse existing parsers?
6) Do the same events generated by different apps or parts of the kernel have the same fields in the same order and in the same format?

All event format problems revolve around these basic goals.

**Comment by rgbriggs** [comment]
On 2018-02-05 04:39, Steve Grubb wrote:
> Normalizing records is good. Always recording information that never
> changes is bad. This wastes disk space, increases processing time,
> consumes more memory, etc. It is OK for records to swing in and out.

This is already a swinging record: PATH

> This is why we have so many auxiliary records. If something is rare,
> it goes in an auxiliary record when it can swing in and out as needed.

Applying this reasoning, an auxilliary record should be added to the
auxilliary PATH record, including appropriate numbering to keep the
"item=" field trackable.

Does this seem reasonable?  That sounds tenuous to me.

Fixing the technical debt in the parser sounds like short term pain for
long term gain.

**Comment by stevegrubb** [comment]
4.5. If neither 3 or 4 is applicable for a new field that needs to be recorded, are we using the shortest, unique field name that describes the purpose of the field? Does it have '_' rather than the preferred '-'
7. Does the value side have any space characters in it causing parsing issues? If the space is required, the field must be encoded. But if its a static field the user has no control over, its preferred to glue the words together with a '-' so as not to waste disk space and increase processing.

**Comment by stevegrubb** [comment]
On Mon, Feb 5, 2018 at 7:52 AM, Richard Guy Briggs <notifications@github.com
> wrote:

> On 2018-02-05 04:39, Steve Grubb wrote:
> > Normalizing records is good. Always recording information that never
> > changes is bad. This wastes disk space, increases processing time,
> > consumes more memory, etc. It is OK for records to swing in and out.
>
> This is already a swinging record: PATH
>

As reported in the first comment, there are only about 20 - 30 files on the
whole system that have capabilities. Since capabilities are rare, they
should not be logged in the general case. The new capability information
adds 50% to the size of the old PATH records.

Reading /usr/include/linux/audit.h, I see:

#define AUDIT_BPRM_FCAPS        1321    /* Information about fcaps
increasing perms */

This is the record that file capabilities belongs in. This way we swing in
the information only when applicable.

> This is why we have so many auxiliary records. If something is rare,
> > it goes in an auxiliary record when it can swing in and out as needed.
>
> Applying this reasoning, an auxilliary record should be added to the
> auxilliary PATH record, including appropriate numbering to keep the
> "item=" field trackable.
>
> Does this seem reasonable? That sounds tenuous to me
>

This does not fix the reported problem which is the record size has
increased about 50% for no additional benefit.


> Fixing the technical debt in the parser sounds like short term pain for
> long term gain.
>

This is not a parser problem. This is why are we recording unchanging
information that increases the record size by a whole lot when there is a
record specifically for this? This is principle #2 above.

**Comment by rgbriggs** [comment]
On 2018-02-05 13:05, Steve Grubb wrote:
> As reported in the first comment, there are only about 20 - 30 files on the
> whole system that have capabilities. Since capabilities are rare, they
> should not be logged in the general case. The new capability information
> adds 50% to the size of the old PATH records.

And if there are two PATH records that each report fcaps, they need
their own auxilliary record each.

> #define AUDIT_BPRM_FCAPS        1321    /* Information about fcaps increasing perms */
> 
> This is the record that file capabilities belongs in. This way we swing in
> the information only when applicable.

This is the record that file capabilities belong in if capabilies
increased perms.

> This is not a parser problem. This is why are we recording unchanging
> information that increases the record size by a whole lot when there is a
> record specifically for this? This is principle #2 above.

The existing auxilliary record isn't exactly for this usage and could
need to be invoked in the same event.

**Comment by stevegrubb** [comment]
On Mon, Feb 5, 2018 at 8:18 AM, Richard Guy Briggs <notifications@github.com
> wrote:

> On 2018-02-05 13:05, Steve Grubb wrote:
> > As reported in the first comment, there are only about 20 - 30 files on
> the
> > whole system that have capabilities. Since capabilities are rare, they
> > should not be logged in the general case. The new capability information
> > adds 50% to the size of the old PATH records.
>
> And if there are two PATH records that each report fcaps, they need
> their own auxilliary record each.
>

Sure.


> > #define AUDIT_BPRM_FCAPS 1321 /* Information about fcaps increasing
> perms */
> >
> > This is the record that file capabilities belongs in. This way we swing
> in
> > the information only when applicable.
>
> This is the record that file capabilities belong in if capabilities
> increased perms.


Why else would a file have capabilities? :-) Files don't have capabilities
without them being used to elevate the privileges. Of course if root
invokes ping, there is no elevation, but then the fs capabilities
privileges didn't matter. If we are installing an updated iputils package,
then recording that it has capabilities is a surprise and should be in a
BPRM_FCAPS record. If we are installing an updated aide package, no file
has capabilities and its not a surprise or newsworthy. IFF we have fs
capabilities, should we consider logging this information in the event.


> > This is not a parser problem. This is why are we recording unchanging
> > information that increases the record size by a whole lot when there is a
> > record specifically for this? This is principle #2 above.
>
> The existing auxilliary record isn't exactly for this usage and could
> need to be invoked in the same event.
>

Such as? Let's talk through this scenario to be sure. If you want to make
an AUDIT_FCAPS record, we could go that route also. But let's be sure
first.

**Comment by pcmoore** [comment]
I have a quick question:

> From @stevegrubb:
> As reported in the first comment, there are only about 20 - 30 files on the
whole system that have capabilities.

I assume this is on some standard config Fedora or RHEL variant, yes?  I wonder about other distributions/solutions as I know the ambient capabilities functionality is something that was wanted/needed because people were using file capabilities and were finding disconnects between their use cases and how things worked.

**Comment by stevegrubb** [comment]
Yes, the above info is based on Fedora/RHEL observations.

**Comment by rgbriggs** [comment]
Some executable with fcaps manipulates two files that both have fcaps.
Example: Some special "mv"-like executable has fcaps and no suid and
we've got a syscall rule to monitor rename.  The source and destination
(that will be unlinked and path reused) files each already have fcaps.
I'd need an AUDIT_SYSCALL record with:
	PROCTITLE
	CWD
	BPRM_FCAPS listing caps of executable
	PATH (item=0) source directory
	PATH (item=1) source file
	PATH_FCAPS (item=1) source file
	PATH (item=2) destination directory
	PATH (item=3) destination file
	PATH_FCAPS (item=3) destination file

**Comment by pcmoore** [comment]
@rgbriggs you lost me on that last comment, I'm sure I'm just missing something so can you elaborate a bit more?

Also, just a reminder to everyone involved in this issue, we've already made the change upstream, so there is a very large barrier to changing it now.  There has been a very disturbing trend recently of wanting to revert/change things that have made it to Linus' tree after weeks/months of review, documentation, and testing; this is not acceptable.

**Comment by rgbriggs** [comment]
On 2018-02-06 12:49, Paul Moore wrote:
> @rgbriggs you lost me on that last comment, I'm sure I'm just missing
> something so can you elaborate a bit more?

The issue is that the existing BPRM_FCAPS record is to document the use
of fcaps to gain permissions to execute a process that it would not have
gained in other ways (eg: suid), while the fcap fields that were
converted to non-swinging were already part of the PATH record (of which
there could be multiple associated to one SYSCALL), each needing its own
auxilliary record to avoid making the PATH auxilliary record longer than
necessary when no fcaps are present (for which I was proposing a new
record type "PATH_FCAPS").  This would be an auxilliary record to an
auxilliary record (yes, you read that correctly).

> Also, just a reminder to everyone involved in this issue, we've
> already made the change upstream, so there is a very large barrier to
> changing it now.  There has been a very disturbing trend recently of
> wanting to revert/change things that have made it to Linus' tree after
> weeks/months of review, documentation, and testing; this is not
> acceptable.

I've already made my postion known on this issue which is why I chose
not to address it in my last comment.

Non-swinging fields, append-only field changes and auxilliary records
are all policies that have been introduced to address inflexibilities in
the parser.

The proposal I presented in my last comment was an example of a way to
try to solve this situation, fully recognizing that it was too late to
use it on this issue due to the offending code already being upstream.

**Comment by pcmoore** [comment]
> From @rgbriggs:
> The issue is that the existing BPRM_FCAPS record is to document the use
of fcaps to gain permissions to execute a process that it would not have
gained in other ways (eg: suid), while the fcap fields that were
converted to non-swinging were already part of the PATH record (of which
there could be multiple associated to one SYSCALL), each needing its own
auxilliary record to avoid making the PATH auxilliary record longer than
necessary when no fcaps are present (for which I was proposing a new
record type "PATH_FCAPS").  This would be an auxilliary record to an
auxilliary record (yes, you read that correctly).

Okay, I think I understand now, thanks for the clarification.  Please don't waste time on these changes right now, it's unlikely I'd accept this for reasons that have already been discussed in this thread.

> From @pcmoore: 
> Also, just a reminder to everyone involved in this issue, we've
 already made the change upstream, so there is a very large barrier to
 changing it now.  There has been a very disturbing trend recently of
 wanting to revert/change things that have made it to Linus' tree after
 weeks/months of review, documentation, and testing; this is not
 acceptable.
>
> From @rgbriggs 
> I've already made my postion known on this issue which is why I chose
not to address it in my last comment.

Yes, that is well understood, and I think we are in agreement here.  My comments were largely aimed at those who have recently started to object to changes only after they have made it into Linus tree after months of review.

**Comment by pcmoore** [comment]
@rgbriggs is recommending we close this as WONTFIX and I agree with that suggestion for all the reasons already discussed.
